## Narative:

- I have been tasked to determine which city has the most flight delays and if there are any carriers who may have an alarming amount of delays. 
  - In order to analyze this data I need to first create a csv file from the table given and then import csv file into Pandas. 
  - Once I have imported data I need to clean the data and transform table to long and tidy format. 
  - Once the data is in proper format I can then use grouping and filtering to analyze data.

In [0]:
import pandas as pd
import numpy as np
import csv

In [0]:
# Assemble data for CSV file
flight_status = [['','','Los Angeles','Phoenix','San Diego','San Francisco','Seattle'],
                  ['ALASKA','on time',497,221,212,503,1841],
                  ['','delayed',62,12,20,102,305],
                  ['','','','','','',''],
                  ['AM West','on time',694,4840,383,320,201],
                  ['','delayed',117,415,65,129,61]]

In [381]:
flight_status

[['', '', 'Los Angeles', 'Phoenix', 'San Diego', 'San Francisco', 'Seattle'],
 ['ALASKA', 'on time', 497, 221, 212, 503, 1841],
 ['', 'delayed', 62, 12, 20, 102, 305],
 ['', '', '', '', '', '', ''],
 ['AM West', 'on time', 694, 4840, 383, 320, 201],
 ['', 'delayed', 117, 415, 65, 129, 61]]

In [357]:
# export data to .csv file
with open('flight_status.csv', 'w') as csvFile:
    writer = csv.writer(csvFile)
    writer.writerows(flight_status)
csvFile.close()

NameError: ignored

In [0]:
# Import data from Github
data = pd.read_csv('https://raw.githubusercontent.com/AVIMARCUS6/DAV-5400/master/Fall/flight_status.csv')

In [383]:
# Drop any row which is fully empty
data=data.dropna(axis=0, how='all', thresh=None, subset=None, inplace=False)
data

Unnamed: 0 Unnamed: 1  Los Angeles  ...  San Diego  San Francisco  Seattle
0     ALASKA    on time        497.0  ...      212.0          503.0   1841.0
1        NaN    delayed         62.0  ...       20.0          102.0    305.0
3    AM West    on time        694.0  ...      383.0          320.0    201.0
4        NaN    delayed        117.0  ...       65.0          129.0     61.0

[4 rows x 7 columns]

In [384]:
# ffill data so the NaNs take on the value of the previous value above
data = data.fillna(method='ffill')
data

Unnamed: 0 Unnamed: 1  Los Angeles  ...  San Diego  San Francisco  Seattle
0     ALASKA    on time        497.0  ...      212.0          503.0   1841.0
1     ALASKA    delayed         62.0  ...       20.0          102.0    305.0
3    AM West    on time        694.0  ...      383.0          320.0    201.0
4    AM West    delayed        117.0  ...       65.0          129.0     61.0

[4 rows x 7 columns]

In [0]:
# Rename unnamed columns to 'Carrier' and 'Status'
data.rename(columns={'Unnamed: 0':'Carrier',
                     'Unnamed: 1':'Status'},
            inplace=True)

In [386]:
# Creates tidy form of data, creating a new column 'City'
data = pd.melt(data, id_vars =['Carrier','Status'], value_vars =['Los Angeles','Phoenix','San Diego','San Francisco','Seattle'],var_name='City') 
data

Carrier   Status           City   value
0    ALASKA  on time    Los Angeles   497.0
1    ALASKA  delayed    Los Angeles    62.0
2   AM West  on time    Los Angeles   694.0
3   AM West  delayed    Los Angeles   117.0
4    ALASKA  on time        Phoenix   221.0
5    ALASKA  delayed        Phoenix    12.0
6   AM West  on time        Phoenix  4840.0
7   AM West  delayed        Phoenix   415.0
8    ALASKA  on time      San Diego   212.0
9    ALASKA  delayed      San Diego    20.0
10  AM West  on time      San Diego   383.0
11  AM West  delayed      San Diego    65.0
12   ALASKA  on time  San Francisco   503.0
13   ALASKA  delayed  San Francisco   102.0
14  AM West  on time  San Francisco   320.0
15  AM West  delayed  San Francisco   129.0
16   ALASKA  on time        Seattle  1841.0
17   ALASKA  delayed        Seattle   305.0
18  AM West  on time        Seattle   201.0
19  AM West  delayed        Seattle    61.0

### Which City has the highest amount of delays?

In [387]:
# Groups data by status, carrier then city
city_delay = data.groupby(['Status','City']).sum()
city_delay

value
Status  City                 
delayed Los Angeles     179.0
        Phoenix         427.0
        San Diego        85.0
        San Francisco   231.0
        Seattle         366.0
on time Los Angeles    1191.0
        Phoenix        5061.0
        San Diego       595.0
        San Francisco   823.0
        Seattle        2042.0

Based on the DataFrame it appears that Phoenix has the highest number of delays, with 427 delays.

***

### From Phoenix, which carrier has the most delays?

In [0]:
# Filters DataFrame to only contain data for Phoenix flights
is_phx = data['City'] == 'Phoenix'

In [0]:
phx = data[is_phx]

In [390]:
# Group new DataFrame of only Phoenix flights by Status and Carrier
phx.groupby(['Status','Carrier']).sum()

value
Status  Carrier        
delayed ALASKA     12.0
        AM West   415.0
on time ALASKA    221.0
        AM West  4840.0

Out of the flights leaving Phoenix, AM West has the most delays with 415 delays, while alaska only has 12 delays.

***

### Changes:

If I were to convert the data back to wide format, I would include the Carrier for each row instead of only one row. I would also switch the status and carrier columns in order to compare flight status against each carrier. 